source/blog/_posts/2015-11-02-javascript-interfaces.md

```diff
@@ -0,0 +1,31 @@
+
+> When I see a bird that walks like a duck, swims like a duck and quacks like a duck, I call that bird a duck.
+
+<img src="/blog/img/dog_quack.jpg" style="width: 190px; float: right; margin: 0 0 30px 40px;" />
+
+It means that an object with the methods ```walk()```, ```swim()``` and ```quack()``` can always be treated as a ```duck```. In JavaScript objects are defined by the methods they implement and not by an explicit type. When you have a second object of type ```dog```, which also implements the ```walk()``` method, it could be treated like a ```duck``` as long as none of the other methods get called. So you never can be sure if a given object has implemented the necessary method until you proof it.
+
+But JavaScript is also a very flexible language and of course there are different ways to implement some kind of interface functionality. You will find many articles around the web about this topic. The problem is, without native support, you always have to manually ensure that a class implements the interface you're providing. So you can mitigate the problem, but there is no realistic way to force a third party developer to use your interface as intended. It is all about providing good documentation to your code.
+
+In Shopware we're using <a href="https://www.sencha.com/products/extjs/" target="_blank">ExtJS</a> for the administration panel of the shop system, which is a really complex JavaScript framework. Ouf course it has features for inheritance like creating and extending classes. But sometimes we're really missing the functionality of an interface, especially when encouraging third party developers to build new modules on top of the existing platform. In the new <a href="https://en.shopware.com/shopware-5-series-digital-publishing/" target="_blank">Digital Publishing</a> module for example, you are able to create custom elements which can be used inside the module. To provide an easy way for creating these elements, we built a kind of abstract class, which defines the basic set of methods and properties the element should use. Developers can extend this abstract class to create new elements right out of the box, without taking care about data handling and all that stuff. But also this is no real interface, so we have to encourage all developers to work through the <a href="https://developers.shopware.com/developers-guide/digital-publishing-elements/" target="_blank">documentation</a> and use the code as intended.
+
+What are your experiences with such a kind of abstraction in JavaScript? Let us know in the comments.
```

Comment:
At the moment we do not have any comment system enabled on the blog. Maybe point them to your twitter handle for comments.